go/bucketpool/bucketpool.go

```diff
@@ -0,0 +1,81 @@
+package bucketpool
+
+import (
+	"sync"
+)
+
+type sizedPool struct {
+	size int
+	pool sync.Pool
+}
+
+func newSizedPool(size int) *sizedPool {
+	return &sizedPool{
+		size: size,
+		pool: sync.Pool{
+			New: func() interface{} { return makeSlicePointer(size) },
+		},
+	}
+}
+
+// Pool is actually multiple pools which store buffers of specific size.
+// i.e. it can be three pools which return buffers 32K, 64K and 128K.
+type Pool struct {
+	maxSize int
+	pools   []*sizedPool
+}
+
+// New returns Pool which has `tiers` buckets with first one returning buffers with `minSize`.
+// Last buffer will have size of minSize * (multiple ^ (tiers - 1)).
+func New(minSize, multiplier, tiers int) *Pool {
+	pools := []*sizedPool{
+		newSizedPool(minSize),
+	}
+	for i := 1; i < tiers; i++ {
+		minSize *= multiplier
+		pools = append(pools, newSizedPool(minSize))
+	}
+	return &Pool{
+		maxSize: minSize,
+		pools:   pools,
+	}
+}
+
+func (p *Pool) findPool(size int) *sizedPool {
+	if size > p.maxSize {
+		return nil
+	}
+	for _, sp := range p.pools {
```

Comment:
Use an expression instead. It will be complex, but efficient (and worth it). This also means that `maxSize` is not needed.
And write lots of tests to prove it's correct, especially boundary conditions :).

Reply:
Ooooh, I got it. Sorry :) Will do.